plugin/folding_range.py

```diff
@@ -0,0 +1,205 @@
+from .core.protocol import FoldingRange
+from .core.protocol import FoldingRangeKind
+from .core.protocol import FoldingRangeParams
+from .core.protocol import Range
+from .core.protocol import Request
+from .core.protocol import UINT_MAX
+from .core.registry import LspTextCommand
+from .core.typing import List, Optional
+from .core.views import range_to_region
+from .core.views import text_document_identifier
+from functools import partial
+import sublime
+
+
+def folding_range_to_range(folding_range: FoldingRange) -> Range:
+    return {
+        'start': {
+            'line': folding_range['startLine'],
+            'character': folding_range.get('startCharacter', UINT_MAX)
+        },
+        'end': {
+            'line': folding_range['endLine'],
+            'character': folding_range.get('endCharacter', UINT_MAX)
+        }
+    }
+
+
+def sorted_folding_ranges(folding_ranges: List[FoldingRange]) -> List[FoldingRange]:
+    # Sort by reversed position and from innermost to outermost (if nested)
+    return sorted(
+        folding_ranges,
+        key=lambda r: (
+            r['startLine'],
+            r.get('startCharacter', UINT_MAX),
+            -r['endLine'],
+            -r.get('endCharacter', UINT_MAX)
+        ),
+        reverse=True
+    )
+
+
+class LspFoldCommand(LspTextCommand):
+    """A command to fold at the current caret position or at a given point.
+
+    Command arguments:
+
+    - `manual`: Should be `true` when invoked from the command palette or via key binding, and `false` for menu items.
```

Comment:
The name of this one feels a little misleading.

Do I understand correctly that you think `manual=False` is appropriate when used within menu item because then `is_visible` and `description` are invoked without triggering the command explicitly? If so then maybe that makes sense but when clicking the menu item it's definitively a "manual" action but it's still invoked with `manual=False`. Would it be clearer if it was just called "from_menu_item"?

Reply:
Yeah, I didn't think much about it and just used "manual" because it was already used in some other commands. I renamed it to "prefetch" and updated the docstring. In practice users should only care about the "strict" argument if they want to add a custom key binding or command palette entry.